paddle/framework/grad_op_desc_maker.h

```diff
@@ -21,27 +23,49 @@ namespace framework {
 
 class GradOpDescMakerBase {
  public:
-  explicit GradOpDescMakerBase(const OpDescBind& fwd_op) : fwd_op_(fwd_op) {}
+  explicit GradOpDescMakerBase(
+      const OpDescBind& fwd_op,
+      const std::unordered_set<std::string>& no_grad_set)
+      : fwd_op_(fwd_op), no_grad_set_(no_grad_set) {}
 
   virtual ~GradOpDescMakerBase() = default;
   virtual std::vector<std::unique_ptr<OpDescBind>> operator()() const = 0;
 
  protected:
-  static std::vector<std::string> ToGradNames(
-      const std::vector<std::string>& var_names) {
+  std::vector<std::string> ToGradNames(
+      const std::vector<std::string>& var_names,
+      bool drop_empty_grad = true) const {
     std::vector<std::string> ret_val;
     ret_val.reserve(var_names.size());
-    std::transform(var_names.begin(), var_names.end(),
-                   std::back_inserter(ret_val), GradVarName);
-    return ret_val;
+    std::transform(
+        var_names.begin(), var_names.end(), std::back_inserter(ret_val),
+        [this](const std::string& fwd_var_name) -> std::string {
+          auto g_name = GradVarName(fwd_var_name);
+          if (this->no_grad_set_.find(g_name) != this->no_grad_set_.end()) {
+            return kEmptyVarName;
+          } else {
+            return g_name;
+          }
+        });
+    if (!drop_empty_grad) {
+      return ret_val;
+    }
+    std::vector<std::string> dropped_ret_val;
+    dropped_ret_val.reserve(ret_val.size());
+    std::copy_if(ret_val.begin(), ret_val.end(),
+                 std::back_inserter(dropped_ret_val),
+                 [](const std::string& str) { return str != kEmptyVarName; });
+    return dropped_ret_val;
```

Comment:
After the filtering, not only output but also input of grad_op can be empty. It's incorrect. Input without gradient should be `zero`, not empty.

Reply:
Done.